facade/src/main/java/org/jboss/pnc/facade/providers/BuildConfigurationProviderImpl.java

```diff
@@ -51,4 +160,165 @@ public BuildConfigurationProviderImpl(BuildConfigurationRepository repository,
+                .validateCondition(!configId.equals(dependencyId), "A build configuration cannot depend on itself")
+                .validateCondition(!dependency.getAllDependencies().contains(buildConfig), "Cannot add dependency from : "
+                        + configId + " to: " + dependencyId + " because it would introduce a cyclic dependency");
+
+        System.out.println("didn't throw any validation errors");
+        buildConfig.addDependency(dependency);
+        repository.save(buildConfig);
+    }
+
+    @Override
+    public void removeDependency(Integer configId, Integer dependencyId) {
+
+        org.jboss.pnc.model.BuildConfiguration buildConfig = repository.queryById(configId);
+        org.jboss.pnc.model.BuildConfiguration dependency = repository.queryById(dependencyId);
+
+        ValidationBuilder.validateObject(buildConfig, WhenUpdating.class)
+                .validateCondition(buildConfig != null, "No build config exists with id: " + configId)
+                .validateCondition(dependency != null, "No dependency build config exists with id: " + dependencyId);
+
+        buildConfig.removeDependency(dependency);
+        repository.save(buildConfig);
+    }
+
+    @Override
+    public void setProductVersion(Integer configId, Integer productVersionId) {
+
+        org.jboss.pnc.model.BuildConfiguration buildConfig = repository.queryById(configId);
+
+        ValidationBuilder.validateObject(buildConfig, WhenUpdating.class)
+                .validateCondition(buildConfig!=null, "No build config exists with id: " + configId);
+
+        ProductVersion productVersion = null;
+
+        if (productVersionId != null) {
+            productVersion = productVersionRepository.queryById(productVersionId);
+        }
+
+        buildConfig.setProductVersion(productVersion);
+        repository.save(buildConfig);
+    }
+
+    @Override
+    public Page<BuildConfiguration> getDependencies(int pageIndex,
+                                                    int pageSize,
+                                                    String sortingRsql,
+                                                    String query,
+                                                    Integer configId) {
+
+        return queryForCollection(pageIndex, pageSize, sortingRsql, query, withDependantConfiguration(configId));
+    }
+
+    @Override
+    public Page<BuildConfigurationRevision> getRevisions(int pageIndex, int pageSize, Integer id) {
+
+        List<BuildConfigurationAudited> auditedBuildConfigs =
+                buildConfigurationAuditedRepository.findAllByIdOrderByRevDesc(id);
+
+        List<BuildConfigurationRevision> toReturn =  nullableStreamOf(auditedBuildConfigs)
+                .map(buildConfigurationRevisionMapper::toDTO)
+                .skip(pageIndex * pageSize)
+                .limit(pageSize)
+                .collect(Collectors.toList());
```

Comment:
While we can do the pagination via the `AuditQuery` (setFirstResult, setMaxResults), I think it's not worth it because when creating the Page object, we need to provide the 'totalHits'. Using AuditQuery, there's no way to tell it that we only want a count of all the audited objects. Instead we can only get the full result list.

Reply:
ok, let's keep it that way and if we decide we don't need total hits/pages we can delete it then